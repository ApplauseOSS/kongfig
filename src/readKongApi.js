import kongState from './kongState';

export default async (adminApi) => {
    return Promise.all([kongState(adminApi), adminApi.fetchPluginSchemas()])
        .then(([state, schemas]) => {
            const prepareConfig = (plugin, config) => stripConfig(config, schemas.get(plugin));
            const parseApiPluginsForSchemes = plugins => parseApiPlugins(plugins, prepareConfig);

            return parseApis(state.apis, parseApiPluginsForSchemes);
        })
};

function parseApis(apis, parseApiPlugins) {
    return apis.map(({
        name, plugins,
        request_host, request_path, strip_request_path, preserve_host, upstream_url,
        id, created_at}) => {
        return {
            name,
            plugins: parseApiPlugins(plugins),
            attributes: {
                request_host,
                request_path,
                strip_request_path,
                preserve_host,
                upstream_url,
            },
            _info: {
                id,
                created_at
            }
        };
    });
}

function parseApiPlugins(plugins, prepareConfig) {
    return plugins.map(({
        name,
        config,
        id, api_id, consumer_id, enabled, created_at
    }) => {
        return {
            name,
            attributes: {
                config: prepareConfig(name, config)
            },
            _info: {
                id,
                //api_id,
                consumer_id,
                enabled,
                created_at
            }
        };
    });
}

function stripConfig(config, schema) {
    const mutableConfig = {...config};

    if (false) {
        // strip default keys
        Object.keys(config).map(key => {
            if (schema[key].hasOwnProperty('default') && schema[key].default === config[key]) {
                delete mutableConfig[key];
            }
        });
    }

    // remove some cache values
    delete mutableConfig['_key_der_cache'];
    delete mutableConfig['_cert_der_cache'];

    return mutableConfig;
}
