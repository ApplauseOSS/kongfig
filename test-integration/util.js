import adminApi from '../src/adminApi';
import readKongApi from '../src/readKongApi';
import execute from '../src/core';

if (typeof process.env.TEST_INTEGRATION_KONG_HOST == 'undefined') {
    throw Error('Please set TEST_INTEGRATION_KONG_HOST env variable');
}

export const testAdminApi = adminApi({
    host: process.env.TEST_INTEGRATION_KONG_HOST,
    https: false,
    ignoreConsumers: false,
    cache: false,
});

const _ignoreKeys = (obj, keys) => {
    if (obj instanceof Array) {
        obj.forEach(function(item) {
            _ignoreKeys(item, keys)
        });
    } else if (typeof obj === 'object') {
        Object.getOwnPropertyNames(obj).forEach((key) => {
            if (keys.indexOf(key) !== -1) {
                obj[key] = '___IGNORED___';
            } else {
                _ignoreKeys(obj[key], keys);
            }
        });
    }

    return obj;
};

export const ignoreKeys = (message, keys) => _ignoreKeys(JSON.parse(JSON.stringify(message)), keys);

export const cleanupKong = async () => {
    console.log("\n\n--- Cleanup \n\n".bold);
    const results = await readKongApi(testAdminApi);
    await execute({
        apis: results.apis.map(api => ({ ...api, ensure: 'removed' })),
        consumers: results.consumers.map(consumer => ({ ...consumer, ensure: 'removed' })),
        plugins: results.plugins.map(plugin => ({ ...plugin, ensure: 'removed' })),
    }, testAdminApi);

    console.log("\n\n--- Cleanup END \n\n".bold);
};
