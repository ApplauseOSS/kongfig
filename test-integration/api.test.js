import execute from '../src/core';
import { testAdminApi, ignoreKeys, cleanupKong } from './util';
import readKongApi from '../src/readKongApi';

let log = [];
const logger = message => {
    log.push(ignoreKeys(message, ['apiId', 'id', 'api_id', 'created_at']));
};

beforeAll(async () => {
    await cleanupKong();
});

afterEach(async () => {
    log = [];
    await cleanupKong();
});

describe("API", () => {
    it("should add an API", async () => {
        const config = {
            apis: [{
                name: "mockbin",
                ensure: "present",
                attributes: {
                    upstream_url: "http://mockbin.com",
                    hosts: ["mockbin.com"]
                }
            }]
        };

        await execute(config, testAdminApi, logger);
        const kongState = await readKongApi(testAdminApi);

        expect(log).toMatchSnapshot();
        expect(ignoreKeys(kongState, ['created_at', 'id'])).toMatchSnapshot();
    });

    it("should add an API with a plugins", async () => {
        const config = {
            apis: [{
                name: "mockbin",
                ensure: "present",
                attributes: {
                    upstream_url: "http://mockbin.com",
                    hosts: ["mockbin.com"]
                },
                plugins: [{
                    name: "key-auth"
                }, {
                    name: "cors",
                    attributes: {
                        config: {
                            credentials: false,
                            preflight_continue: false,
                            max_age: 7000
                        }
                    }
                }]
            }]
        };

        await execute(config, testAdminApi, logger);
        const kongState = await readKongApi(testAdminApi);

        expect(log).toMatchSnapshot();
        expect(ignoreKeys(kongState, ['created_at', 'id'])).toMatchSnapshot();
    });
});
