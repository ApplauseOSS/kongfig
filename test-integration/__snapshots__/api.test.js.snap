// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API should add an API 1`] = `
Array [
  Object {
    "params": Object {
      "body": Object {
        "hosts": Array [
          "mockbin.com",
        ],
        "name": "mockbin",
        "upstream_url": "http://mockbin.com",
      },
      "endpoint": Object {
        "name": "apis",
      },
      "method": "POST",
    },
    "type": "action",
  },
  Object {
    "content": Object {
      "created_at": "___IGNORED___",
      "hosts": Array [
        "mockbin.com",
      ],
      "http_if_terminated": true,
      "https_only": false,
      "id": "___IGNORED___",
      "name": "mockbin",
      "preserve_host": false,
      "retries": 5,
      "strip_uri": true,
      "upstream_connect_timeout": 60000,
      "upstream_read_timeout": 60000,
      "upstream_send_timeout": 60000,
      "upstream_url": "http://mockbin.com",
    },
    "type": "response-ok",
  },
]
`;

exports[`API should add an API 2`] = `
Object {
  "apis": Array [
    Object {
      "_info": Object {
        "created_at": "___IGNORED___",
        "id": "___IGNORED___",
      },
      "attributes": Object {
        "hosts": Array [
          "mockbin.com",
        ],
        "http_if_terminated": true,
        "https_only": false,
        "preserve_host": false,
        "retries": 5,
        "strip_uri": true,
        "upstream_connect_timeout": 60000,
        "upstream_read_timeout": 60000,
        "upstream_send_timeout": 60000,
        "upstream_url": "http://mockbin.com",
      },
      "name": "mockbin",
      "plugins": Array [],
    },
  ],
  "consumers": Array [],
  "plugins": Array [],
}
`;

exports[`API should add an API with a plugins 1`] = `
Array [
  Object {
    "params": Object {
      "body": Object {
        "hosts": Array [
          "mockbin.com",
        ],
        "name": "mockbin",
        "upstream_url": "http://mockbin.com",
      },
      "endpoint": Object {
        "name": "apis",
      },
      "method": "POST",
    },
    "type": "action",
  },
  Object {
    "content": Object {
      "created_at": "___IGNORED___",
      "hosts": Array [
        "mockbin.com",
      ],
      "http_if_terminated": true,
      "https_only": false,
      "id": "___IGNORED___",
      "name": "mockbin",
      "preserve_host": false,
      "retries": 5,
      "strip_uri": true,
      "upstream_connect_timeout": 60000,
      "upstream_read_timeout": 60000,
      "upstream_send_timeout": 60000,
      "upstream_url": "http://mockbin.com",
    },
    "type": "response-ok",
  },
  Object {
    "params": Object {
      "body": Object {
        "config": Object {
          "key_names": Array [
            "foobar",
          ],
        },
        "name": "key-auth",
      },
      "endpoint": Object {
        "name": "api-plugins",
        "params": Object {
          "apiId": "___IGNORED___",
          "pluginName": "key-auth",
        },
      },
      "method": "POST",
    },
    "type": "action",
  },
  Object {
    "content": Object {
      "api_id": "___IGNORED___",
      "config": Object {
        "anonymous": "",
        "hide_credentials": false,
        "key_in_body": false,
        "key_names": Array [
          "foobar",
        ],
      },
      "created_at": "___IGNORED___",
      "enabled": true,
      "id": "___IGNORED___",
      "name": "key-auth",
    },
    "type": "response-ok",
  },
  Object {
    "params": Object {
      "body": Object {
        "config": Object {
          "credentials": false,
          "max_age": 7000,
          "preflight_continue": false,
        },
        "name": "cors",
      },
      "endpoint": Object {
        "name": "api-plugins",
        "params": Object {
          "apiId": "___IGNORED___",
          "pluginName": "cors",
        },
      },
      "method": "POST",
    },
    "type": "action",
  },
  Object {
    "content": Object {
      "api_id": "___IGNORED___",
      "config": Object {
        "credentials": false,
        "max_age": 7000,
        "preflight_continue": false,
      },
      "created_at": "___IGNORED___",
      "enabled": true,
      "id": "___IGNORED___",
      "name": "cors",
    },
    "type": "response-ok",
  },
]
`;

exports[`API should add an API with a plugins 2`] = `
Object {
  "apis": Array [
    Object {
      "_info": Object {
        "created_at": "___IGNORED___",
        "id": "___IGNORED___",
      },
      "attributes": Object {
        "hosts": Array [
          "mockbin.com",
        ],
        "http_if_terminated": true,
        "https_only": false,
        "preserve_host": false,
        "retries": 5,
        "strip_uri": true,
        "upstream_connect_timeout": 60000,
        "upstream_read_timeout": 60000,
        "upstream_send_timeout": 60000,
        "upstream_url": "http://mockbin.com",
      },
      "name": "mockbin",
      "plugins": Array [
        Object {
          "_info": Object {
            "created_at": "___IGNORED___",
            "id": "___IGNORED___",
          },
          "attributes": Object {
            "config": Object {
              "anonymous": "",
              "hide_credentials": false,
              "key_in_body": false,
              "key_names": Array [
                "foobar",
              ],
            },
            "enabled": true,
          },
          "name": "key-auth",
        },
        Object {
          "_info": Object {
            "created_at": "___IGNORED___",
            "id": "___IGNORED___",
          },
          "attributes": Object {
            "config": Object {
              "credentials": false,
              "max_age": 7000,
              "preflight_continue": false,
            },
            "enabled": true,
          },
          "name": "cors",
        },
      ],
    },
  ],
  "consumers": Array [],
  "plugins": Array [],
}
`;
